### PR TITLE
Asad change password form revision

### PR DESCRIPTION
**Pull request by asad041**
1. public->index.php [route (a)] *
2. app->controller->usercontroller [settingaction(u), changepasswordsubmit(a)] *
3. app->model->auth [checkpasswordbyid(a)] *
4. app->model->form->changepassword.php (a) *
5. app->model->users.php [saveuser(u)] *
6. app->view->user->settings.phtml (u) *
7. app->view->user->change-password.phtml (a) *

[u: updated, a: added, d: delete, p:progress, *:done]

### CODE REVIEW

File: app/Model/Form/changePassword.php [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+     */
+    public function validate(array $params)
+    {
+        parent::validate($params);
+
+        $isValidnewPassword = $this->validator->isValidPasswordString($params['newPassword']);
+
+        if ($isValidnewPassword) {
+            $isSamePassword = $params['currentPassword'] == $params['newPassword'];
+            if($isSamePassword){
```
Comment (thorie7912) [inline app/Model/Form/changePassword.php]: Let's run the `cs-fix` alias to ensure the formatting of this file.

File: app/Model/Form/changePassword.php [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    private $session;
+
+    
+    public function __construct(Validator $validator, Users $users, Auth $auth, Session $session)
+    {
+        $this->defineFields(['currentPassword', 'newPassword', 'confirmPassword']);
+        $this->validator = $validator;
+        $this->users = $users;
+        $this->auth = $auth;
+        $this->session = $session;
```
Comment (thorie7912) [inline app/Model/Form/changePassword.php]: Is session used in this class?

File: app/Model/Users.php [hunk 3]
```diff
@@ -86,23 +86,28 @@ public function loadUser(User $user)
...
         $params = [];
         $typesMap = ['email' => 's', 'verifyCode' => 's', 'password' => 's', 'passwordResetCode' => 's'];
         $types = [];
         
         foreach ($user->getModifiedProperties() as $property) {
-            $set[] = "$property = ?";
-            $types[] = $typesMap[$property];
+            
+            if($property == 'id'){
+                // nothing to add
```
Comment (thorie7912) [inline app/Model/Users.php]: I think `id` shouldn't be a modified property, so the above `getModifiedProperties` shouldn't return `id`. So, do we still need this check?

File: app/Model/Form/changePassword.php [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+                $isConfirmPassword = $params['newPassword'] == $params['confirmPassword'];
+                if(!$isConfirmPassword){
+                    $this->setError('confirmPassword', 'Password not matched');
+                }
+            }
+        }else{
+            $this->setError('newPassword', 'New Password must be between ' . Validator::MIN_PASSWORD_LENGTH .' and ' . Validator::MAX_PASSWORD_LENGTH . ' characters.');
+        }
+
+        $isValidCurrPassword = $this->validator->isValidPasswordString($params['currentPassword']) && $this->auth->checkPasswordById($this->session->get('userId'), $params['currentPassword']);
```
Comment (thorie7912) [inline app/Model/Form/changePassword.php]: It looks like `isValidPasswordString` is already checked above. Can we reuse that result?

File: app/Controller/UserController.php [hunk 5]
```diff
@@ -343,6 +352,26 @@ public function passwordResetSubmitAction(array $p)
...
+            return;
+        }
+
+        $user = $this->createUserObject();
+        $user->id = $this->session->get('userId');
+        $user->password = $p['newPassword'];
+
+        $this->users->saveUser($user);
+        
+        $this->session->set('message', 'Successfully password changed');
```
Comment (thorie7912) [inline app/Controller/UserController.php]: Message should be `Successfully changed password`

File: app/Model/Form/changePassword.php [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    public function validate(array $params)
+    {
+        parent::validate($params);
+
+        $isValidnewPassword = $this->validator->isValidPasswordString($params['newPassword']);
+
+        if ($isValidnewPassword) {
+            $isSamePassword = $params['currentPassword'] == $params['newPassword'];
+            if($isSamePassword){
+                $this->setError('newPassword', 'New Password must be different than old.');
```
Comment (thorie7912) [inline app/Model/Form/changePassword.php]: Message should be `New password should not match your existing password.`

File: app/Model/Form/changePassword.php [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+            }else{
+                $isConfirmPassword = $params['newPassword'] == $params['confirmPassword'];
+                if(!$isConfirmPassword){
+                    $this->setError('confirmPassword', 'Password not matched');
```
Comment (thorie7912) [inline app/Model/Form/changePassword.php]: Message should be `Passwords do not match, please try again.`

File: app/Model/Form/changePassword.php [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+                }
+            }
+        }else{
+            $this->setError('newPassword', 'New Password must be between ' . Validator::MIN_PASSWORD_LENGTH .' and ' . Validator::MAX_PASSWORD_LENGTH . ' characters.');
+        }
+
+        $isValidCurrPassword = $this->validator->isValidPasswordString($params['currentPassword']) && $this->auth->checkPasswordById($this->session->get('userId'), $params['currentPassword']);
+
+        if(!$isValidCurrPassword){
+            $this->setError('currentPassword', 'Incorrect Password.');
```
Comment (thorie7912) [inline app/Model/Form/changePassword.php]: Message should be `Existing password is incorrect, please try again.`

File: app/Model/Form/changePassword.php [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+use App\Model\Auth;
+use App\Model\Users;
+use App\Util\Validator;
+use App\Util\Session;
+
+/**
+ * @property $currentPassword
+ * @property $newPassword
+ */
+class ChangePassword extends AbstractForm
```
Comment (thorie7912) [inline app/Model/Form/changePassword.php]: File name should be uppercase `ChangePassword.php` to match the class here.

File: app/Model/Form/changePassword.php [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+     * will be updated to what was submitted.
+     *
+     * @param array $params
+     * @throws \Exception
+     */
+    public function validate(array $params)
+    {
+        parent::validate($params);
+
+        $isValidnewPassword = $this->validator->isValidPasswordString($params['newPassword']);
```
Comment (thorie7912) [inline app/Model/Form/changePassword.php]: Camel-case should be consistent `$isValidNewPassword`. It is fine to keep `$params['newPassword']`.

File: app/View/user/change-password-form.phtml [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+      </div>
+      <div class="control-group">
+        <label class="control-label" for="inputNewPassword">New password</label>
+        <div class="controls">
+          <input type="password" class="form-control" id="inputNewPassword" name="newPassword" placeholder="New password" value="">
+          <div style='color:red'><?=$vars['formErrors']['newPassword']?></div>
+        </div>
+      </div>
+      <div class="control-group">
+        <label class="control-label" for="inputConfirmPassword">Confirm password</label>
```
Comment (thorie7912) [inline app/View/user/change-password-form.phtml]: To be more clear, let's use the label `Confirm new password`

File: app/Model/Auth.php [hunk 12]
```diff
@@ -29,4 +29,21 @@ public function checkPassword($email, $password)
...
+
+    public function checkPasswordById($id, $password)
+    {
+        sleep(1);
+        $hash = password_hash($password, PASSWORD_BCRYPT);
+        $query = 'select password from users where id = ?';
+        $params = [
+            'id' => $id,
+        ];
+        $result = $this->mysql->query($query, 's', $params);
```
Comment (thorie7912) [inline app/Model/Auth.php]: Let's use `i` instead of `s` because the id should be an integer.

File: app/Model/Users.php (hunk 3 above)
Comment (asad041) [reply app/Model/Users.php]: yeah it shouldn't but in my case it returned

Comment (thorie7912) [reply app/Model/Users.php]: Oh, I think it returned because in the controller there is:
```
$user->id = $this->session->get('userId');
```
If you change that to `$user->init(['id' => $this->session->get('userId')])` then it won't consider `id` to be modified, because it won't be changed from `null` to the number. It's a little weird, but it makes it easier to save the data later.

File: app/Model/Users.php [hunk 13]
```diff
@@ -102,7 +102,7 @@ public function saveUser(User $user)
             
             if ($property == 'password') {
                 $params[] = password_hash($user->password, PASSWORD_BCRYPT);
-            } else {
+            } else if($property != 'id') {
```
Comment (thorie7912) [inline app/Model/Users.php]: Looks like you can revert all the changes in this file.

**Comment by thorie7912** [comment]
Please add unit tests for the following:

* Controller\UserController::passwordChangeSubmitAction
* Model\Auth::checkPasswordById
* Model\Form\ChangePassword::validate

Thanks!

**Comment by asad041** [comment]
I did try but not work out.... getting error

File: tests/app/Model/Form/WithChangePasswordExpectations.php [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+        ->with('pass123')
+        ->willReturn(true);
+    }
+    
+    private function expectAuthCheckPasswordByIdPass()
+    {
+        $this->mocks[Session::class]->expects($this->once())
+        ->method('get')
+        ->with('userId')
+        ->willReturn(3);
```
Comment (thorie7912) [inline tests/app/Model/Form/WithChangePasswordExpectations.php]: Let's move this into it's own method `expectSessionGetUserId()` and call it from the test function.

File: app/Controller/UserController.php [hunk 15]
```diff
@@ -343,6 +352,26 @@ public function passwordResetSubmitAction(array $p)
         $this->successfulLogin($p['email']);
     }
     
+    public function passwordChangeSubmitAction(array $p)
+    {   
+        $this->changePasswordForm->validate($p);
```
Comment (thorie7912) [inline app/Controller/UserController.php]: Missing initialization for this form variable in constructor.

File: app/Controller/UserController.php [hunk 16]
```diff
@@ -343,6 +352,26 @@ public function passwordResetSubmitAction(array $p)
...
+
+        if ($this->changePasswordForm->hasErrors()) {
+            $this->session->set('changePasswordFormState', $this->changePasswordForm->getState());
+            $this->headers->redirect('user/settings');
+            return;
+        }
+
+        $user = $this->createUserObject();
+        $user->id = $user->init(['id' => $this->session->get('userId')]);
```
[the remark this answers is not in the document]
Comment (thorie7912) [reply app/Controller/UserController.php]: the `$user` object will have the `id` property assigned automatically due to the init(...) function call
if you also assign `$user->id = ...` then you're basically setting it **twice**

**Comment by thorie7912** [comment]
Next task is to update tests/integration.sh
Your goal is to write a simple test that validates the functionality using curl commands, closer to a user's perspective.
Let me know if it doesn't work. You might have to install cygwin.

File: tests/integration.sh [hunk 17]
```diff
@@ -16,15 +16,15 @@ check() { grep -ie "$1" out.html > /dev/null; if [ $? = 1 ]; then echo ""; echo
...
-post() { rm -f out.html; echo ""; echo "Submitting form with $2 to: $1"; curl -c cookies.txt -L -s -k -o out.html -d $2 $1; alwayscheck; }
+browse() { rm -f out.html; echo ""; echo "Browsing to: $1"; curl -c cookies.txt -L -s -o out.html $1; alwayscheck; }
+post() { rm -f out.html; echo ""; echo "Submitting form with $2 to: $1"; curl -c cookies.txt -L -s -o out.html -d $2 $1; alwayscheck; }
 cleanup() { rm -f out.html; rm cookies.txt; }
 
 # -------------------------------------------------
 # Test cases
 # -------------------------------------------------
 
-rand=`php -r 'echo rand(0,99999999);'`
```
Comment (thorie7912) [inline tests/integration.sh]: I think you are not using git correctly here. How are you pushing your changes? Are you using some kind of GUI or are you using git on the command line?

File: app/Model/Users.php [hunk 18]
```diff
@@ -86,11 +86,11 @@ public function loadUser(User $user)
     }
     
     public function saveUser(User $user)
-    {
+     {
         if (!isset($user->id)) {
             throw new \Exception('User ID required to save');
         }
-        
+
```
Comment (thorie7912) [inline app/Model/Users.php]: You can revert the changes in this file completely, by grabbing the `master` branch version into yours.
Run this command:

`git checkout master app/Model/users.php`

Then, add and commit it.

`git add app/Model/Users.php`
`git commit -m "Restoring master version of Users.php"`
`git push origin HEAD`

File: tests/app/Model/Form/ChangePasswordTest.php [hunk 19]
```diff
@@ -0,0 +1,90 @@
...
+        $this->assertEquals($this->expectedErrors, $this->form->getErrors());
+    }
+
+    private function assertValidateFailure()
+    {
+        $this->assertTrue($this->form->hasErrors());
+        $this->assertEquals('', $this->form->currentPassword);
+        $this->assertEquals('', $this->form->newPassword);
+        $this->assertEquals('', $this->form->confirmPassword);
+        //$this->assertEquals($this->expectedErrors, $this->form->getErrors());
```
Comment (thorie7912) [inline tests/app/Model/Form/ChangePasswordTest.php]: Not needed?

File: tests/app/Model/Form/LoginTest.php [hunk 20]
```diff
@@ -1,97 +1,97 @@
...
+    }
+    
+    private function assertValidationFailure()
+    {
+        $this->assertTrue($this->form->hasErrors());
+        $this->assertEquals('foo@bar.com', $this->form->email);
+        $this->assertEquals('', $this->form->password);
+        $this->assertEquals($this->expectedErrors, $this->form->getErrors());
+    }
+}
```
Comment (thorie7912) [inline tests/app/Model/Form/LoginTest.php]: I think we can revert the changes in this file using the same procedure.

File: tests/app/Model/Form/WithLoginExpectations.php [hunk 21]
```diff
@@ -1,58 +1,58 @@
...
+    }
+    
+    private function expectUtilValidatorFailInvalidPasswordChars()
+    {
+        $this->setMockExpectations(Validator::class, [
+            'isEmailChars' => [$this->once(), ['foo@bar.com'], true],
+            'isPasswordChars' => [$this->once(), ['somepass123'], false],
+        ]);
+    }
+
```
Comment (thorie7912) [inline tests/app/Model/Form/WithLoginExpectations.php]: Probably can revert these changes too.

File: tests/app/Model/Form/ChangePasswordTest.php [hunk 22]
```diff
@@ -0,0 +1,90 @@
...
```
Comment (thorie7912) [inline tests/app/Model/Form/ChangePasswordTest.php]: Is this needed?

File: tests/app/Model/Form/WithChangePasswordExpectations.php [hunk 23]
```diff
@@ -0,0 +1,63 @@
...
+            Auth::class,
+            Session::class
+        ]);
+    }
+    
+    private function expectUtilValidatorPass()
+    {
+        //$this->setMockExpectations(Validator::class, [
+        //    'isValidPasswordString' => [$this->at(0), ['pass123'], true],
+        //]);
```
Comment (thorie7912) [inline tests/app/Model/Form/WithChangePasswordExpectations.php]: Is this needed?

**Comment by thorie7912** [comment]
Coverage is increasing...
```
\App\Model\Form::ChangePassword
  Methods:  50.00% ( 1/ 2)   Lines:  83.33% ( 20/ 24)
```
Let's get it to 100%

File: app/Model/Users.php (hunk 18 above)
Comment (asad041) [reply app/Model/Users.php]: did that but same result

File: tests/integration.sh [hunk 24]
```diff
@@ -53,16 +53,19 @@ signupAndLogin()
...
 
-orders()
+changePassword()
 {
-    browse "$URL/order/history"
+    browse "$URL/user/settings"
     check 'button'
+
+    post "$URL/user/password-change-submit" "currentPassword=pass100&newPassword=pass123&confirmPassword=pass123";
+    check ''
```
Comment (thorie7912) [inline tests/integration.sh]: we should check something here, to make sure it worked ... some piece of success text that is unique to the change password

File: app/View/user/change-password-form.phtml [hunk 25]
```diff
@@ -0,0 +1,39 @@
...
+        </div>
+      </div>
+      <div class="control-group" style="margin-top: 5px">
+        <div class="controls">
+          <button type="submit" class="btn btn-primary">Update password</button>
+        </div>
+      </div>
+    </form>
+  </div>
+</div>
```
Comment (thorie7912) [inline app/View/user/change-password-form.phtml]: This rendering looks like this:
image
I'm wondering if we can line up the label with the input like this:
image

File: app/View/user/change-password-form.phtml [hunk 26]
```diff
@@ -0,0 +1,39 @@
...
+    <p>
+        Enter a new password below.
+    </p>
+</div>
+
+<div class="row">
+  <div class="col-md-6">
+    <form class="form-horizontal" method="post" action="<?=$vars['baseUrl']?>/user/password-change-submit">
+      <div class="control-group">
+        <label class="control-label" for="inputCurrentPassword">Current Password</label>
```
Comment (thorie7912) [inline app/View/user/change-password-form.phtml]: To be consistent with the other labels lowercase, let's use: `Current password`

File: app/View/user/change-password-form.phtml (hunk 25 above)
Comment (asad041) [reply app/View/user/change-password-form.phtml]: done :+1:
